src/librbd/ObjectMap.cc

```diff
@@ -138,8 +138,7 @@ bool ObjectMap<I>::update_required(const ceph::BitVector<2>::Iterator& it,
   ceph_assert(ceph_mutex_is_locked(m_lock));
   uint8_t state = *it;
   if ((state == new_state) ||
-      (new_state == OBJECT_PENDING && state == OBJECT_NONEXISTENT) ||
-      (new_state == OBJECT_NONEXISTENT && state != OBJECT_PENDING)) {
+      (new_state == OBJECT_PENDING && state == OBJECT_NONEXISTENT)) {
```

Comment:
I am not sure about this.

Note, `update_required` was introduced in e5b4188635c4ee1ee0c4353cfc5ecd6e887d536b to avoid unnecessary map updates for discard operation. Before that change we did on discard:
```
update_map OBJECT_NONEXISTENT -> OBJECT_PENDING
write_object
update_map OBJECT_PENDING -> OBJECT_NONEXISTENT
```
and with the change the `update_map` operations are skipped for this particular (discard) case.

Now with your change, as I understand it, it will not skip the second update, which will  be a considerable performance degradation for certain scenarios (see the tracker ticket from the commit above).

Another thing, `update_required` seems to be used only for head (`snap_id == CEPH_NOSNAP`) while you probably need to make it work for a snapshot case too.